pulser/simulation/simresults.py

```diff
@@ -14,63 +14,320 @@
+
+    def get_final_state(self) -> qutip.Qobj:
+        """Get the final state of the simulation as a diagonal density matrix.
+
+        Note: This is not the density matrix of the system, but is a convenient
+            way of computing expectation values of observables.
+
+        Returns:
+            qutip.Qobj: States probability distribution as a density matrix.
+        """
+        return self.get_state(self.sim_times[-1])
+
+    def expect(self, obs_list: Sequence[Union[qutip.Qobj, ArrayLike]]
+               ) -> list[Union[float, complex, ArrayLike]]:
+        """Calculates the expectation value of a list of observables.
+
+        Args:
+            obs_list (array-like of qutip.Qobj or array-like of numpy.ndarray):
+                A list of observables whose expectation value will be
+                calculated. If necessary, each member will be transformed into
+                a qutip.Qobj instance.
+
+        Note: This only works for diagonal observables, since results have been
+            projected onto the Z basis.
+
+        Returns:
+            list: the list of expectation values of each operator.
+        """
+        for obs in obs_list:
+            if not isdiagonal(obs):
+                raise ValueError(f"Observable {obs} is non-diagonal.")
+
+        return super().expect(obs_list)
+
+    def _calc_weights(self, t: float) -> ArrayLike:
+        N = self._size
+        bitstrings = [np.binary_repr(k, N) for k in range(2**N)]
+        t_index = self._get_index_from_time(t)
+        return [self._results[t_index][b] for b in bitstrings]
+
+    def _standard_dev(self, op: qutip.Qobj) -> ArrayLike:
+        """Returns the square root of the variance of operator op."""
+        return cast(ArrayLike,
+                    np.sqrt(qutip.variance(op, self.states) / self.N_measures)
+                    )
+
+    def _get_error_bars(self, op: qutip.Qobj) -> Tuple[ArrayLike, ArrayLike]:
+        moy = self.expect([op])[0]
+        st = self._standard_dev(op)
+        return moy, st
+
+    def plot(self, op: qutip.Qobj, fmt: str = '.',
+             label: str = '', error_bars: bool = True) -> None:
+        """Plots the expectation value of a given operator op.
+
+        Note: The observable must be diagonal.
+
+        Args:
+            op (qutip.Qobj): Operator whose expectation value is wanted.
+            error_bars (bool): Choose to display error bars.
+            fmt (str): Curve plot format.
+            label (str): y-Axis label.
+        """
+        if not isdiagonal(op):
+            raise ValueError(f"Observable {op} is non-diagonal.")
+        if error_bars:
+            moy, st = self._get_error_bars(op)
+            plt.errorbar(self.sim_times, moy, st, fmt=fmt, lw=1, capsize=3,
+                         label=label)
+        else:
+            plt.plot(self.sim_times, self.expect([op])[0], fmt,
+                     label=label)
+        plt.xlabel('Time (µs)')
+        plt.ylabel('Expectation value')
```

Comment:
Wouldn't this work if you moved `plot` in `CleanResults` to `SimulationResults`?
```suggestion
        if error_bars:
            moy, st = self._get_error_bars(op)
            plt.errorbar(self.sim_times, moy, st, fmt=fmt, lw=1, capsize=3,
                         label=label)
            plt.xlabel('Time (µs)')
            plt.ylabel('Expectation value')
        else:
            super().plot(op, fmt=fmt, label=label, error_bars=error_bars)
```

Reply:
Yes, this would work, I just didn't want to change the `CleanResults.plot()` signature by adding an `error_bars` arg. I'll do this though, maybe it's cleaner

Reply:
Oh, I understand now, that's a pretty good way to do it ! It's not changing the signature. I'll do it